Feature: select+joins analysis with pandas DataFrames

il-ia-ni opened the pull request:
Hey Phillip :)

This is my first commit to the branch for testing pandas dataframes with the SQL Alchemy lists of select+join-results.

Right now there seem to be problems with filling DataFrames with data, even though the sqlalchemy lists are correctly parsed to DataFrames. This should have something to do with the detached status of the lists upo closing the Session.

Here are the links to the topic: 
1. Error: [Parent instance <x> is not bound to a Session; (lazy load/deferred load/refresh/etc.) operation cannot proceed](https://docs.sqlalchemy.org/en/14/errors.html#error-bhk3)
2. State Management of SQL Alchemy query results (only for ORM-instances): [https://docs.sqlalchemy.org/en/14/orm/session_state_management.html#session-object-states](https://docs.sqlalchemy.org/en/14/orm/session_state_management.html#session-object-states)

Commit:
Added comments to the sessionmaker in progress

Commit:
Merge branch 'main' into feature-move_sessions_2_sessionmaker

Commit:
upd loguru rotation for custom debug logs

Commit:
Update debug_formatter.py

Commit:
Update debug_formatter.py

Commit:
Update debug_formatter.py

Commit:
upd simple_select methods to support global session instances

Commit:
Upd select_joins methods to support global session instances

Commit:
Tried adding a seccond logging formatter for lower-lvl logging

Commit:
Refactor adding/deleting new instances logic. added TODOs

Commit:
Update main.py

Commit:
Added severity level filet for the custom loguru formatter

Commit:
revised loguru severities after filtering of format

Commit:
Update creating_tables.py

Commit:
small upds for working with the branch

Commit:
unimportant fix

Commit:
Added logic to generate DataFrames from lists of ORM- and Engine.Row-Instances

Dataframes get no data as the underlying instances are detached upon closing session within the methods!

https://docs.sqlalchemy.org/en/14/errors.html#error-bhk3

https://docs.sqlalchemy.org/en/14/orm/session_state_management.html#session-object-states

Commit:
Added dataframe data extractions

Commit:
Update main.py

Commit:
Upd select_joins with core statements

Commit:
Extended attrs list for selections with join

Commit:
upd docstrings

Commit:
upd select_joins orm method

Commit:
Merge branch 'main' into feature-joins-analysis-w-pandas

Commit:
moved testinmg of adding and deleting objs to adding_data.py

Commit:
restored pandas testing in main

Phillip-Oliver commented on main.py:
```diff
@@ -1,17 +1,18 @@
 
 debug_format()
-url = build_full_url(url_SQLServerTestDBMS + SQLServerTestDBs.MASTER_DB.value + "?driver=ODBC+Driver+17+for+SQL+Server")
+url = build_full_url(url=url_SQLServerTestDBMS + SQLServerTestDBs.MASTER_DB.value, odbc_driver="ODBC Driver 17 for SQL Server")
```
Here we have build_full_url and make_engine but but in sql_server_engine we created another engine called: engine_SQLServerTest_MainDB. Does it make sense to create one session in main ?

il-ia-ni replied:
Updated ;) The engine is now created in sql_server_engine script by calling a method from the db_connection. The main-script now only imports the ready-to-use engine!

Commit:
Update main.py

Commit:
Added axis names to the test dataframes

Commit:
Merge branch 'main' into feature-joins-analysis-w-pandas

Commit:
Upd method for creating engines + refactored sql server instance

Commit:
refactored code to the engine changes

Phillip-Oliver commented on ddl_scripts/adding_data.py:
```diff
@@ -43,4 +44,21 @@ def add_new_objs(session: Session, orm_obj: Base):
+logger.info(f"Fetching the following test obj: {test_obj1}")  # TODO: trace severity requires a custom formatter
+
+test_obj2 = global_session.get(SignalMeta, "test_obj_2")
+logger.info(f"Fetching the following test obj: {test_obj2}")  # TODO: trace severity requires a custom formatter
```
Das könnten wir ideal in einer loop machen oder ? Vllt so etwas hier: 
signal_meta = [('test_obj_1', 'Test signal meta', 'testCaster', 'test_category', 'Test_testCaster', 'TestUnit', 'This is a test object') ('test_obj_2', 'Test signal meta 2', 'testCaster 2', 'test_category 2', 'Test_testCaster2 ', 'TestUnit 2', 'This is a test object 2'), ...]

Deine add_all[] erlaubt ja schon Listen ;)

Aber das ist nur als Hinweis oder Idee zu verstehen ;)

il-ia-ni replied:
Ich habe diesen Kommentar abgearbeitet! Die Methode ist nun mit dem "Splat"-Operator in der Parameterliste (*test_objs) erweitert. Eine echt coole Funktion! 

[Hier ](https://github.com/il-ia-ni/python_sql_tutorial/pull/4/commits/478b379216feb031d78f61ae2b51f15320f205f3)ist der Commit dazu.

il-ia-ni replied:
Und hier ist der Artikel zu dem Splat-Operator *args: [https://realpython.com/python-kwargs-and-args/](https://realpython.com/python-kwargs-and-args/)

Phillip-Oliver commented on dql_scripts/select_joins.py:
```diff
@@ -2,22 +2,43 @@
+)
+
+
+def get_select_join_rowslist_result(engine: sqlalchemy.engine, select_stmt):
+    with engine.connect() as connection:
```
Wir hatten ja festgestellt, dass es der bessere Weg ist über session.execute() zu gehen. Dann kannst du überlegen, ob du das hier löschen möchtest oder nicht ? ;) Überlasse ich aber dir !

il-ia-ni replied:
Das ist nur hier für die Informationszwecke zu der Core API :) Ansonsten mache ich bereits alle Tests mit DataFrames ausschließlich mit der ORM API and anhand einer Session und der get_select_join_orm_result()

Phillip-Oliver reviewed (commented):
Hier bin ich mir nicht sicher: 
![grafik](https://user-images.githubusercontent.com/54893338/218402101-74deeb7b-c4db-40e4-a8f1-1dff0a6148ed.png)

Die Funktion ist ja sicherlich zu Testzwecken gedacht. Aber wir müssen nochmal schauen/recherchieren, ob es Standard ist, in einer Funktion drei mal session.execute() aufzurufen. Und da wir kein session.add_all() verwenden, frage ich mich,ob das session.commit() notwendig ist.

Phillip-Oliver commented on main.py:
```diff
@@ -23,24 +21,52 @@
-# jnt_sel.get_select_join_orm_result(global_session, jnt_sel.select_join_orm_stmt1)  # Join with select.join_from()
-# jnt_sel.get_select_join_orm_result(global_session, jnt_sel.select_join_orm_stmt2)  # Join with select.join()
-# jnt_sel.get_select_join_orm_result(global_session, jnt_sel.select_join_orm_stmt3)  # Join with join() and explicit ON
+joins_scalar = jnt_sel.get_select_join_orm_result(global_session, jnt_sel.select_join_orm_stmt1)  # Join with select.join_from()
+joins_rows = jnt_sel.get_select_join_rowslist_result(sqlserver_engine, jnt_sel.select_join_core_stmt1)  # Join with select.join_from()
```
Hier verwenden wir einmal eine session und einmal eine Engine ? Hat das einen Grund ? Kommen da unterschiedliche Ergebnisse bei raus ?

il-ia-ni replied:
Das hier war auch nur zur Prüfung geschrieben, die Ergebnisse sind 100% identisch. Also ich lösche das bald weg, wenn ich ein Skript für pandas anlege.

il-ia-ni replied:
Beide Anweisungen sind nun zu Testzwecken in das Dataframes.py Skript verschoben!

Phillip-Oliver commented on main.py:
```diff
@@ -23,24 +21,52 @@
+logger.info("A DataFrame with following parameters was created from the rows list: \n", rows_df.info())
+logger.info(rows_df.head(5))
+
+pa_query_df = pa.read_sql_query(
+    # See https://pandas.pydata.org/docs/reference/api/pandas.read_sql_query.html#pandas.read_sql_query
+    sql=dql_scripts.select_joins.select_join_orm_stmt1,
+    con=sqlserver_engine,
+    dtype={"signal_id": "string", "name": "string", "model_name": "string"},  # otherwise parsed as dtype 'object'
+    # parse_dates={"update_date": "%c"},
+    # parse_dates={"update_date": {"utc": True, "format": "%c"}}
+)
```
Finde ich gut, dass wir das betrachtet haben, können wir ja im weiteren Verlauf noch berücksichtigen !

il-ia-ni replied:
Ich habe schon fast vergessen, dass pandas immer alles als Objekte anstatt Strings parsen :D Es war auf jeden Fall eine gute Erinnerung daran!

Commit:
Upd proj to function with new firewall limitations

Commit:
upd SQL Server enigine params for pandas compability

Commit:
Added multiple filter conditions in with-statements

Commit:
Added a script with methods for creating dataframes

Commit:
formatted main.py

Commit:
Added todo for unit tests for future prs

Commit:
Updated adding_data method with the splat operator to support any amount of test_objs

https://realpython.com/python-kwargs-and-args/

il-ia-ni commented:
> Hier bin ich mir nicht sicher: ![grafik](https://user-images.githubusercontent.com/54893338/218402101-74deeb7b-c4db-40e4-a8f1-1dff0a6148ed.png)
> 
> Die Funktion ist ja sicherlich zu Testzwecken gedacht. Aber wir müssen nochmal schauen/recherchieren, ob es Standard ist, in einer Funktion drei mal session.execute() aufzurufen. Und da wir kein session.add_all() verwenden, frage ich mich,ob das session.commit() notwendig ist.

Bin mir auch hierzu nicht so ganz sicher! Da SQL Alchemy in der Dokumentation den "Commit as you go" Weg immer nimmt (so schreiben sie es selber), habe ich es auch in meiner Testmethode übernommen. Sonst muss ich noch rausfinden, wie man diesen .begin() auf eine Session-Instanz anwendet...

Hier ist was in der [SQL Alchemy Dokumentation](https://docs.sqlalchemy.org/en/14/tutorial/dbapi_transactions.html#committing-changes) zu dem Unterschied zwischen den "Commit as you go"- und "Begin once"-Commitprinzipien steht:

_“Begin once” style is often preferred as it is more succinct and indicates the intention of the entire block up front. However, within this tutorial we will normally use “commit as you go” style as it is more flexible for demonstration purposes._

Commit:
Upd table definitions to the main scheme of the test DB

Commit:
Added refactored defect_event_root_cause script

Package: data_analysis

Commit:
Added own defect_root_cause_pairs script

To the data analysis package

Fixed the problem with binding non-scalar tuple as a parameter of a text query. https://stackoverflow.com/a/56382828

Commit:
Moved test lines to dataframes.py and upd main.py

il-ia-ni commented:
Hey @Phillip-Oliver , ich habe heute alle Neuerungen hochgeladen! 
Unter anderem sind nun alle meinen Methoden für Erstellung von DataFrames in die Package data_analysis migriert worden.

Ich habe auch deine beiden Skripts (defect_event_root_cause und defect_root_cause_pairs) auch im data_analysis Package überarbeitet und ein Paar interessanten Sachen da rausgefunden.

Auch meine Test Datenbank ist nun komplett dem Schema mit dem falschen Namen losgeworden, alle Tabellen, Prozeduren und Funktionen sind nun unter dem Schema "Main" zu finden: Dies ermöglicht eigentlich, diese PR zu schließen und dann kann ich die nächste PR mit der lokalen fallback-SQLite DB fertig machen!

Also bitte nochmals um deine Kommentare :)

Commit:
Update defect_root_cause_pairs.py

Commit:
Update creating_tables.py

Commit:
Update dataSources.xml

Commit:
Create sshConfigs.xml

Commit:
Update creating_tables.py

Commit:
Update dataframes.py

Commit:
Update creating_tables.py

Commit:
Update main.py

Commit:
Upd scripts with testing of group creation from jupiter notebook

Commit:
Added matplotlib to the project

Commit:
Added testing of plotting func to the main script

Phillip-Oliver commented on data_analysis/dataframes.py:
```diff
@@ -0,0 +1,105 @@
+from db_engines.sql_server_engine import engine_sqlservertest_main as sqlserver_engine
+
+
+def create_df_from_list(rows_data_list: list, columns_list: list) -> pd.DataFrame:
+    """
+    Creates a pandas DataFrame based on a list of selection result Rows instances that were already created using f.e.
+    SQL Alchemy Core API or ORM API
+    :param rows_data_list: a list of Rows instances of a select result
+    :param columns_list: an iterable with the strings of attributes names of the select result
+    :return: a pandas dataframe with the data of the list with the selection result
+    """
+    result_df = pd.DataFrame(rows_data_list, columns=columns_list)
+    logger.debug("A DataFrame with following parameters was created from the list: \n", result_df.info())
+
+    return result_df
```
Für die Umwandlung einer Liste in einen DF extra eine Funktion ? ;)

il-ia-ni replied:
Tja danke für den Hinweis 😆 Mir ist während der Verbesserung des Codes irgendwie komplett entlaufen, dass ich hier was anderes machen wollte... Meine Idee war eigentlich noch eine statische Typ-Prüfung einzuführen. Ansonsten macht die Methode gar keinen Sinn.

Ich habe sie jetzt mit Union Type Annotations erweitert, damit nur Rows von SQL Alchemy akzeptiert werden. [Hier ist der Commit ](https://github.com/il-ia-ni/python_sql_tutorial/pull/4/commits/7619049e3e5a30808c6adff4f55b56fa48b0f7bb) dazu. Auch den Methodennamen habe ich angepasst. 

Theoretisch könnte man weitere Validierungen von Rows vor der Erstellung von DataFrames einführen. Es war auf jeden Fall eine gute Übung zu den Type Annotations in Python 😉

Phillip-Oliver commented on data_analysis/defect_root_cause_pairs.py:
```diff
@@ -0,0 +1,105 @@
+# non-scalar variables (Sets, Tuples, Lists can be added to text queries can be added using expanding bind parameters:
+# https://stackoverflow.com/a/56382828
+# https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.text
+create_root_cause_groups_text_clause = text(
+    "WITH A AS ("
+    "   SELECT A.create_date, A.event_id, B.signal_id, A.behaviour_pattern_id, A.strand_id, A.slab_id, B.importance "
+    "   FROM main.defect_event A INNER JOIN main.defect_root_cause B "
+    "       on A.event_id=B.event_id "
+    "   WHERE A.event_id IN ("
+    "       SELECT event_id "
+    "       FROM main.defect_root_cause "
+    "       WHERE signal_id NOT in :signals "
+    "           AND slab_id IS NOT null "
+    "           AND create_date BETWEEN CAST(:start_date AS DATE) AND GETDATE() "
+    "       GROUP BY event_id HAVING COUNT(event_id) > 1 AND COUNT(event_id) < 3)"
+    "   )"
+    "SELECT "
+    "   create_date, "
+    "   event_id, "
+    "   behaviour_pattern_id, "
+    "   strand_id, "
+    "   slab_id, "
+    "   importance,"
+    "   STUFF((SELECT ',' + CAST(signal_id AS varchar(1000)) "
+    "       FROM A AS innerTable "
+    "       WHERE innerTable.event_id = p.event_id "
+    "       FOR XML PATH('')),1,1,'') "
+    "AS signal_pattern "
+    "FROM A AS p "
+    "GROUP BY p.event_id, create_date, behaviour_pattern_id, strand_id, slab_id, importance")
+
+
+def root_cause_pairs_query(date: str, signals: tuple[str, ...]) -> str:
+    # tuple[str] vs tuple[str, ...] see: https://stackoverflow.com/questions/72001132/python-typing-tuplestr-vs-tuplestr
+
+    # f"" is formatted string literal. An overview of Python String formatting: https://realpython.com/python-f-strings/
+    # f""" """ allows writing a multiline string with no escape character \ before each new line
```
Und was von beidem funktioniert besser ? :)

il-ia-ni replied:
Auf jeden Fall sind[ die f-Strings](https://realpython.com/python-f-strings/) die beste Lösung! 😄 
Das einzige, worauf man aufpassen muss ist dass die Strings manchmal von SQL Alchemy in integers umgewandelt werden, auch wenn es eigentlich ein Datum ist... Das habe ich für die Zeile mit der skalaren Variable "date" wie folgt gelöst:
`AND create_date BETWEEN CAST(\'{date}' AS DATE) and GETDATE()`

Das ist aber immer noch 1000 Mal einfacher, als für text-Konstrukte [die ganzen Binding-Methoden für nicht-skalare Variablen](https://stackoverflow.com/a/56382828) benutzen zu müssen... War aber auf jeden Fall der Übung wert 📌

Phillip-Oliver commented on data_analysis/defect_root_cause_pairs.py:
```diff
@@ -0,0 +1,105 @@
+    # text_statement = create_root_cause_groups_text_clause.bindparams(bindparam('signals', expanding=True))
+
+    # Option 2:
+    query = root_cause_pairs_query(date=date, signals=signals_to_filter)
+    with session as s:
+        # result = s.execute(text_statement, params={'signals': signals_to_filter, 'start_date': date})
+        result = s.execute(query)
```
Wäre es vllt. auch möglich einen Filter für caster einzubauen? 
Fällt mir gerade so ein !

Die caster wären '2c' oder '1'

il-ia-ni replied:
Ich habe nun den Filter im **extract_root_cause_pairs** Skript eingeführt. [Hier ist der Commit](https://github.com/il-ia-ni/python_sql_tutorial/pull/4/commits/27efbdff02987ed9ee817b62ed43fe94cc6c3784) dazu.

Es gibt jedoch ein ziemlich großes Problem mit den SQLAlchemy Statements in Form der f-Strings: Ein Tuple in Python darf nur dann einen Wert haben, wenn hinter ihm ein Komma steht. Solchen Tuple schafft aber SQL nicht in der "WHERE IN ('1', )"-Anweisung zu erledigen. 

Ich habe deswegen TODOs reingeschrieben, da ansonsten man doppelt einen denselben Wert reingeben muss (siehe die Zeile 128) 😱

Phillip-Oliver commented on main.py:
```diff
@@ -1,20 +1,27 @@
-global_engine = make_engine(url, 1000)
-global_session = get_session(global_engine)
+debug_format()  # adds a custom format for debug-level of loguru (saving local .log-files)
+global_session = get_session(sqlserver_engine)
```
Ich glaube diese plotting Sachen würde ich in ein extra modul packen, also nicht in main.py.

Hast du das mal mit Testdaten ausprobiert ob das läuft ?

il-ia-ni replied:
Yep, ich habe nun das **Main.py** vom Testcode befreit und ein **plots.py** Skript für die Matplotlib-basierten Methoden erstellt. Die Methode fürs Gruppieren der Signaldaten von Events habe ich ins **dataframes.py** Skript verschoben. Hier ist [der Kommit dazu](https://github.com/il-ia-ni/python_sql_tutorial/pull/4/commits/544fecc22138acc324ffe49bf6093e8f235ece52).

Es sind nun auch DocString zu jeder Methode drin sowie ein Paar To-Dos 😃 

Leider kann ich keine Screenshots mehr auf dem Master Client auf GutHub hochladen... 🙄 Nach dem ich die Datanbank restrukturiert habe, habe ich tatsächlich mit ein Paar klinDensätzten signal_data aus der KundenDB bezogen und Plots erstellt! Es klappt alles gut 🏅

Phillip-Oliver reviewed (approved):
Stabile Arbeit !
Kannst ja mal über die Kommentare rüber gehen. Aber ich denke, wir können es bald mergen ;)

Commit:
Refactored create_df_from_list function

Uses Union Type Annotations now: https://docs.python.org/3.10/whatsnew/3.10.html#pep-604-new-type-union-operator
and
https://stackoverflow.com/questions/38854282/do-union-types-actually-exist-in-python#38854368

Commit:
Created plots script, added docstrings to signals groupping and plotting methods

Commit:
FIX: Upd parsing JSON signal_data with ORM cls of SQL Alchemy

Commit:
removed todo

Commit:
Added caster filter for extract_root_cause_pairs + TODO

Single value-tuple does not with f-Strings SQL Queries that contain WITH IN!

Commit:
Added TODO to the last code from jupiter notebook

Commit:
cleaned up the main script
